src/riot-rs-threads/src/threadlist.rs

```diff
@@ -5,22 +5,31 @@ use crate::{ThreadId, ThreadState, THREADS};
 /// Manages blocked [`super::Thread`]s for a resource, and triggering the scheduler when needed.
 #[derive(Debug, Default)]
 pub struct ThreadList {
-    /// Next thread to run once the resource is available.
-    pub head: Option<ThreadId>,
+    /// Last thread that was added to the list.
+    ///
+    /// `Thread.thread_blocklist` is circular, therefore the next thread for
+    /// the tail is the head of the list.
+    pub tail: Option<ThreadId>,
 }
 
 impl ThreadList {
     /// Creates a new empty [`ThreadList`]
     pub const fn new() -> Self {
-        Self { head: None }
+        Self { tail: None }
     }
 
     /// Puts the current (blocked) thread into this [`ThreadList`] and triggers the scheduler.
     pub fn put_current(&mut self, cs: CriticalSection, state: ThreadState) {
         THREADS.with_mut_cs(cs, |mut threads| {
-            let thread_id = threads.current_thread.unwrap();
-            threads.thread_blocklist[usize::from(thread_id)] = self.head;
-            self.head = Some(thread_id);
+            let thread_id = threads.current_pid().unwrap();
+            if let Some(tail) = self.tail {
+                let head = threads.thread_blocklist[usize::from(tail)];
```

Comment:
Unfortunately the compiler can't interfere the type if we just use `into()` because multiple types implement `SliceIndex`. Even if we in practice only have implemented `Into<usize>` for `ThreadId`.